man/scummtr.1
.\" SPDX-License-Identifier: MIT
.\"
.\" Copyright (c) 2020-2025 Donovan Watteau
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a copy
.\" of this software and associated documentation files (the "Software"), to deal
.\" in the Software without restriction, including without limitation the rights
.\" to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
.\" copies of the Software, and to permit persons to whom the Software is
.\" furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
.\" AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
.\" OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
.\" THE SOFTWARE.
.Dd $Mdocdate: April 18 2025 $
.Dt SCUMMTR 1
.Os
.Sh NAME
.Nm scummtr
.Nd import and export fan translations for SCUMM engine games
.Sh SYNOPSIS
.Nm scummtr
.Fl i
.Op Fl bcqvw
.Op Fl l Ar language
.Fl g Ar gameid
.Fl p Ar gamedir
.Fl f Ar input
.Nm scummtr
.Fl o
.Op Fl bchHInqvw
.Op Fl a Ar restypes
.Op Fl A Ar restypes
.Op Fl l Ar language
.Fl g Ar gameid
.Fl p Ar gamedir
.Fl f Ar output
.Nm scummtr
.Fl L
.Sh DESCRIPTION
The
.Nm
tool imports and exports text for SCUMM engine games.
The mode of operation is chosen with one of the following options:
.Bl -tag -width Ds
.It Fl i
Import text into the game data files.
.It Fl o
Export text from the game data files.
.It Fl L
List all supported games with their respective
.Ar gameid .
.El
.Pp
The other options are as follows:
.Bl -tag -width Dslanguage
.It Fl a Ar restypes
Protect the specified SCUMM resource types with extra padding, when
exporting.
.Pp
Some game resources are created with a short name and use a longer
name later in the game.
In such cases,
.Ql @
padding characters must be added to prevent game errors when the resource
is renamed (this is especially true with the original interpreters).
The
.Fl a
option does this action automatically.
.Pp
Valid resource types for the
.Fl a
option are:
.Bl -tag -width Ds
.It Ic a
Pad actor resources.
.It Ic o
Pad object resources.
.It Ic v
Pad verb resources.
.El
.It Fl A Ar restypes
Same as
.Fl a ,
with variable IDs taken into account.
This protects even more resources, at the cost of a slight memory increase
which is negligible nowadays.
.Pp
Using
.Fl A Ar aov
by default is recommended.
.It Fl b
Translation file will be handled in binary mode, instead of text.
.Pp
Note that this makes most other options inoperative.
This flag may not correctly work with all games.
.It Fl c
Ease the use of some Western European characters, by interpreting
the text in the Windows-1252 encoding.
If neither
.Fl c
nor
.Fl r
is used,
non-ASCII characters will be handled as escape sequences.
.Pp
Note that you can only use the characters available in
your current SCUMM fonts.
.It Fl f Ar file
The path to the translation file (default:
.Pa scummtr.txt ) .
.It Fl g Ar gameid
The ID of the game variant to be translated, as given by
.Fl L .
.It Fl h
Include a small header at the start of each line, indicating the
script/block context of the current string.
.Pp
It starts with the room number (e.g.\&
.Dq "001:" ) ,
then the SCUMM object type (e.g.\&
.Dq "ONv2" ) ,
and finally the unique ID of that object (e.g.\&
.Dq "#0394" ) .
.It Fl H
Represent escape sequences as hexadecimal codes, instead of decimal codes.
.It Fl I
Include the current SCUMM instruction opcode, before each line (e.g.\&
.Dq "(D8)"
for that opcode).
.Pp
If the current string is not part of an opcode,
.Dq "(__)"
will be used.
.It Fl l Ar language
V1 and V2 games only (Maniac Mansion and Zak McKracken).
Choose the language charset to be used by
.Fl c .
.Pp
Possible values are:
.Bl -tag -width Ds
.It Ic de
Original German charset.
.It Ic en
Original English charset (default).
.It Ic fr
Original French charset.
.It Ic it
Original Italian charset.
.El
.It Fl n
Never include any
.Dq ";; ScummTR note:"
comment.
.Pp
These comments are added by default to help some text editors infer
the proper encoding of the translation files, but they may interfere with
some derivative tools which do not expect them.
.Fl n
restores the original ScummTR 0.4.0 behavior of not emitting those lines.
.It Fl p Ar gamedir
The path to the game (default: current directory).
.It Fl q
Quiet mode.
Suppress informational output.
.It Fl r
Raw text.
Preserve the original text encoding of the game.
.Pp
This disables any text encoding conversion and escaping:
the text is interpreted
.Dq as-is
(except for non-printable ASCII characters and some internal
SCUMM sequences, which remain escaped),
and you must configure your text editor to recognize
the intended encoding (usually an older MS-DOS code page).
.Pp
This can be useful for non-Western-European languages.
.It Fl v
Verbose mode.
.It Fl w
Use Windows newline characters (CRLF).
This is usually compatible with more text editors, and thus recommended.
.Pp
Note: if using this option, it's important to use it
.Em both
for output
.Pq Fl o
and input
.Pq Fl i
calls.
Otherwise, bogus
.Dq \e013
would be added at the end of each game string, possibly causing issues.
ScummTR 0.6.0 tries to detect such cases, but there is no guarantee.
.El
.Sh EXAMPLES
Extract the text of the original Monkey Island 2 game to a
Windows-1252 file, with added context and protected resource names:
.Pp
.Dl $ scummtr -g monkey2 -cwh -A aov -p /path/to/MI2 -of mi2_orig.txt
.Pp
Import a new French translation into the game files:
.Pp
.Dl $ scummtr -g monkey2 -cwh -A aov -p /path/to/MI2 -if mi2_fr.txt
.Pp
Extract the text of the Japanese version of Monkey Island 2 (FM-TOWNS)
in Shift_JIS, from the current directory:
.Pp
.Dl $ scummtr -g monkey2 -rw -of mi2_towns_jpn.txt
.Pp
Export the text of a German Zak McKracken V2 game, padding resources,
with default paths:
.Pp
.Dl $ scummtr -g zakv2 -l de -cw -A aov -o
.Sh HISTORY
The
.Nm
tool was written between 2003 and 2005
by Thomas Combeleran for the ATP team,
and was open-sourced in 2020 under the MIT license.
.Sh CAVEATS
Your text editor must not remove trailing spaces or empty lines,
otherwise you will have import errors.
.Pp
Since each game and each variant of each game had its own list of
accepted characters (usually a subset of an MS-DOS code page),
it is not possible to know if a character is available without
testing it in the game.
See also
.Xr scummfont 1
in order to visualize or modify the included font tables.
.Pp
Unicode files are not supported, and non-Latin-script alphabets
need extra work for convenient use.
.Pp
On case-sensitive file systems, game data files must currently
follow the DOS convention of being all-uppercase to be properly
detected.


NEWS.md
## ScummTR 0.6.0 (202?-??-??)

### Bugfixes

- ScummTR: Fix a fatal "ERROR: Unknown function id 0x2E" error when trying to import/export text for the German releases of Indy3. This is because of an original issue in the official German releases of the game, where some Eszett characters were badly encoded.

### New features

- ScummTR: try to detect and warn when lines end with bogus `\013` sequences, which can cause some subtle issues with ScummVM or the original interpreters. These sequences would appear when the `-w` option was used for an initial export, but forgotten in a subsequent import.
- ScummTR: when doing an import, try to detect files with Windows CRLF end of lines, so that the `-w` option will be turned on, even if the user forgot to do so. (It's still recommended to explicitly add this option when working with such files, though.)
- ScummTR: add a new `-n` option, which prevents ScummTR from emitting any `;; ScummTR note:` comment. This may help some older derivative tools, such as ScummSpeaks or LanguageBundleBuilder, which don't expect them (since original ScummTR 0.4 never produced them).
- Various manual page and warning message improvements.

### Developer notes

- The project can now be built with Link-Time Optimizations (LTO), by giving the `ENABLE_LTO=ON` option to CMake (if your environment supports it).

## ScummTR 0.5.1 (2022-02-27)

### Bugfixes

- ScummTR/ScummRP: Fix a fatal "Bad OIv2 offset" error with (at least) some versions of Maniac Mansion, Zak McKracken and Indy3. This regression was introduced in 0.5.0.

### Known bugs

- ScummTR/ScummRP: Mark all V2 versions of Maniac Mansion as unsupported, and not just the English V2 version. The French V2 version has the same corruption problem as the English version, for example.

## ScummTR 0.5.0 (2022-02-22)

### New features

- MS-DOS binaries have been added, on top of the macOS, Linux and Windows binaries.
- macOS binaries now have support for PowerPC and Apple silicon processors.
- Manual pages have been added (with .txt versions for systems where the original format is impractical).
- ScummTR: add a new `-r` option, which interprets the text of the game "as-is", in its original encoding. This can be useful for non-Latin languages such as Japanese.
- ScummTR: make it possible to import back a translation file created with the `-h` and/or `-I` options.
- ScummTR: add some encoding headers at the start of the translation file, so that most text editors will properly detect its encoding.
- ScummTR/ScummRP: make "XX.LFL should actually end at 0x1234" a warning which is always displayed, instead of being only printed in verbose mode.
- ScummFont: when importing a font, you now don't have to manually replace the original file with the `-new` file anymore.
- ScummFont: provide more details if an unsupported BMP file is given as an input.

### Bugfixes

- ScummTR/ScummRP: Fix a game corruption bug with at least LOOM-EGA (and possibly some earlier games too), such as clicking on the hole of the leftmost tree of the forest at the start of the game. In the official pre-built binaries, only `scummtr-0.4.2-win32.zip` was impacted by this corruption issue.
- ScummTR/ScummRP: fix a fatal "Duplicate offset in index" error with some versions of LOOM-EGA-EN.
- ScummTR/ScummRP: don't fail when the DISK09.LEC file is missing for MONKEY1-EGA, since it was only available through the Roland Update.
- ScummTR/ScummRP: don't fail and report missing files for the 4-disk floppy versions of MONKEY1, because only the 8-disk floppy versions need them.
- ScummTR/ScummRP: detect and reject the default version of MONKEY1-EGA that Limited Run Games sold in its Monkey Island Anthology; its DISK04.LEC file is known to be corrupted, but we have a guide describing how to recover it.

src/ScummTr/scummtr.cpp
/*
 * SPDX-License-Identifier: MIT
 *
 * Copyright (c) 2003-2005 Thomas Combeleran
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * Note: the last "original" binary was built on: 2003-10-12 13:54:10.
 * Various changes were made in 2004 and in 2005, but were never released,
 * except maybe to some translation team members. Look for the SCUMMTR_
 * #ifdef's to learn more about the changes.
 */

#include "common/backup.hpp"
#include "common/io.hpp"
#include "common/toolbox.hpp"

#include "scummtr.hpp"
#include "trblock.hpp"

#include <cstring>

/*
 * ScummTr
 */

const char *const ScummTr::NAME = "ScummTr";
const char *const ScummTr::VERSION = "0.6.0";
const char *const ScummTr::AUTHOR = "Thomas Combeleran";

const ScummRp::Parameter ScummTr::_trParameters[] =
{
	{ 'g', ScummRp::_paramGameId, sizeof ScummRp::_paramGameId, false },
	{ 'f', ScummTr::_paramTextFile, sizeof ScummTr::_paramTextFile, true },
	{ 'l', ScummTr::_paramLanguage, sizeof ScummTr::_paramLanguage, false },
	{ 'p', ScummRp::_paramGameDir, sizeof ScummRp::_paramGameDir, true },
	{ 'a', ScummTr::_paramPaddedRsc, sizeof ScummTr::_paramPaddedRsc, false },
	{ 'A', ScummTr::_paramPaddedRsc, sizeof ScummTr::_paramPaddedRsc, false },
	{ '\0', nullptr, 0, false }
};

char ScummTr::_paramTextFile[512] = "scummtr.txt";
char ScummTr::_paramLanguage[3] = "en";
char ScummTr::_paramPaddedRsc[16] = "";

int ScummTr::_textOptions = Text::TXT_NULL;
int ScummTr::_paddedRsc = 0;

ScummTr::RscNameLimits ScummTr::_rscNameLimits = { { 0, 0, 0 } };
bool ScummTr::_exportWithPadding = false;
bool ScummTr::_maxPadding = false;

void ScummTr::_explore(TreeBlock &tree, int action, Text &text)
{
	TreeBlockPtr blockPtr;

	tree.firstBlock();
	while ((blockPtr = tree.nextBlock()) != nullptr)
	{
		if (blockPtr.is<LFLFPack>() || blockPtr.is<RoomBlock>())
		{
			ScummTr::_explore(*blockPtr, action, text);
		}
		else
		{
			switch (blockPtr->getTag())
			{
			case MKTAG4('L','E','C','F'):
			case MKTAG2('L','E'):
				{
					TreeBlockPtr lecf;

					lecf = new LECFPack(*blockPtr);
					ScummTr::_explore(*lecf, action, text);
				}
				break;
			case MKTAG4('S','C','R','P'):
			case MKTAG2('S','C'):
			case MKTAG4('S','C','v','3'):
			case MKTAG4('S','C','v','2'):
			case MKTAG4('S','C','v','1'):
			case MKTAG4('E','N','C','D'):
			case MKTAG4('E','X','C','D'):
			case MKTAG2('E','N'):
			case MKTAG2('E','X'):
			case MKTAG4('E','N','v','3'):
			case MKTAG4('E','X','v','3'):
			case MKTAG4('E','N','v','2'):
			case MKTAG4('E','X','v','2'):
			case MKTAG4('E','N','v','1'):
			case MKTAG4('E','X','v','1'):
			case MKTAG4('L','S','v','3'):
				if (action == ScummRp::ACT_IMPORT)
					ScriptBlock(*blockPtr).importText(text);
				else if (action == ScummRp::ACT_EXPORT)
					ScriptBlock(*blockPtr).exportText(text, ScummTr::_exportWithPadding);
				else if (action == ScummTr::ACT_RSCNAMELIMITS)
					ScriptBlock(*blockPtr).getRscNameLimits();
				break;
			case MKTAG4('L','S','C','R'):
			case MKTAG2('L','S'):
				if (action == ScummRp::ACT_IMPORT)
					ScriptBlock(*blockPtr, (ScummRp::_game.version == 7) ? 2 : 1).importText(text);
				else if (action == ScummRp::ACT_EXPORT)
					ScriptBlock(*blockPtr, (ScummRp::_game.version == 7) ? 2 : 1).exportText(text, ScummTr::_exportWithPadding);
				else if (action == ScummTr::ACT_RSCNAMELIMITS)
					ScriptBlock(*blockPtr, (ScummRp::_game.version == 7) ? 2 : 1).getRscNameLimits();
				break;
			case MKTAG4('O','B','C','D'):
				ScummTr::_explore(*blockPtr, action, text);
				break;
			case MKTAG4('O','B','N','A'):
				if (action == ScummRp::ACT_IMPORT)
					ObjectNameBlock(*blockPtr).importText(text);
				else if (action == ScummRp::ACT_EXPORT)
					ObjectNameBlock(*blockPtr).exportText(text, ScummTr::_exportWithPadding);
				else if (action == ScummTr::ACT_RSCNAMELIMITS)
					ObjectNameBlock(*blockPtr).getRscNameLimits();
				break;
			case MKTAG4('V','E','R','B'):
				if (action == ScummRp::ACT_IMPORT)
					ObjectCodeBlock(*blockPtr).importText(text);
				else if (action == ScummRp::ACT_EXPORT)
					ObjectCodeBlock(*blockPtr).exportText(text, ScummTr::_exportWithPadding);
				else if (action == ScummTr::ACT_RSCNAMELIMITS)
					ObjectCodeBlock(*blockPtr).getRscNameLimits();
				break;
			case MKTAG2('O','C'):
			case MKTAG4('O','C','v','3'):
				if (action == ScummRp::ACT_IMPORT)
					OldObjectCodeBlock(*blockPtr).importText(text);
				else if (action == ScummRp::ACT_EXPORT)
					OldObjectCodeBlock(*blockPtr).exportText(text, ScummTr::_exportWithPadding);
				else if (action == ScummTr::ACT_RSCNAMELIMITS)
					OldObjectCodeBlock(*blockPtr).getRscNameLimits();
				break;
			case MKTAG4('O','C','v','2'):
			case MKTAG4('O','C','v','1'):
				if (action == ScummRp::ACT_IMPORT)
					OldObjectCodeBlockV1(*blockPtr).importText(text);
				else if (action == ScummRp::ACT_EXPORT)
					OldObjectCodeBlockV1(*blockPtr).exportText(text, ScummTr::_exportWithPadding);
				else if (action == ScummTr::ACT_RSCNAMELIMITS)
					OldObjectCodeBlockV1(*blockPtr).getRscNameLimits();
				break;
			default:
				break;
			}
		}
	}
}

#ifdef SCUMMTR_HAS_GOOD_GCC_DIAGNOSTIC_PRAGMA_FEATURES
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wformat"
#pragma GCC diagnostic ignored "-Wformat-nonliteral"
#pragma GCC diagnostic ignored "-Wformat-security"
#endif
void ScummTr::_processGameFilesV123()
{
	char dataFileName[32];
	std::string indexPath(ScummRp::_paramGameDir);
	TreeBlockPtr index;
	Text::Charset charset;

	charset = ScummTr::_selectCharset();
	indexPath += '/';
	indexPath += ScummRp::_game.indexFileName;
	index = ScummRp::_newIndex(indexPath.c_str());
	ScummRp::_exploreIndex<ScummRp::ACT_LOAD>(*index);
	{
		Text text(ScummTr::_paramTextFile, ScummTr::_textOptions, charset);

		if (ScummRp::_options & ScummRp::OPT_EXPORT)
			text.addExportHeaders();

		if (ScummRp::_options & ScummRp::OPT_IMPORT)
			text.ensureNoCrlfMisuse();

		for (int i = 1; i < 98; ++i)
		{
			std::string dataPath(ScummRp::_paramGameDir);
			TreeBlockPtr room;

			snprintf(dataFileName, sizeof(dataFileName), ScummRp::_game.dataFileName, i); // ignore -Wformat-security here, ScummRp::_game.dataFileName is internal and safe
			dataPath += '/';
			dataPath += dataFileName;
			if (File::exists(dataPath.c_str()))
			{
				room = ScummRp::_newLFL(dataPath.c_str(), i);
				if (ScummRp::_options & ScummRp::OPT_IMPORT)
				{
					ScummTr::_explore(*room, ScummRp::ACT_IMPORT, text);
				}
				else
				{
					if (ScummTr::_exportWithPadding)
					{
						ScummIO::setQuiet(true);
						ScummTr::_explore(*room, ScummTr::ACT_RSCNAMELIMITS, text);
						ScummIO::setQuiet(false);
					}
					ScummTr::_explore(*room, ScummRp::ACT_EXPORT, text);
				}
			}
		}
	}

	if (ScummRp::_options & ScummRp::OPT_IMPORT)
	{
		ScummIO::setQuiet(true);
		ScummRp::_exploreIndex<ScummRp::ACT_SAVE>(*index);
		ScummIO::setQuiet(false);
	}
}

void ScummTr::_processGameFilesV4567()
{
	char dataFileName[32];
	std::string indexPath(ScummRp::_paramGameDir);
	TreeBlockPtr index;
	int numberOfDisks;
	Text::Charset charset;

	charset = ScummTr::_selectCharset();
	indexPath += '/';
	indexPath += ScummRp::_game.indexFileName;
	index = ScummRp::_newIndex(indexPath.c_str());
	ScummRp::_exploreIndex<ScummRp::ACT_LOAD>(*index);
	numberOfDisks = ScummRp::_mainTocSet.roomToc.numberOfDisks();
	ScummRp::_prepareTmpIndex();
	{
		Text text(ScummTr::_paramTextFile, ScummTr::_textOptions, charset);

		if (ScummRp::_options & ScummRp::OPT_EXPORT)
			text.addExportHeaders();

		if (ScummRp::_options & ScummRp::OPT_IMPORT)
			text.ensureNoCrlfMisuse();

		for (int i = 1; i < numberOfDisks; ++i)
		{
			std::string dataPath(ScummRp::_paramGameDir);
			TreeBlockPtr disk;

			snprintf(dataFileName, sizeof(dataFileName), ScummRp::_game.dataFileName, i); // ignore -Wformat-security here, ScummRp::_game.dataFileName is internal and safe
			dataPath += '/';
			dataPath += dataFileName;
			disk = new BlocksFile(dataPath.c_str(), ScummRp::_fileOptions, ScummRp::_backupSystem, i, MKTAG4('D','I','S','K'), ScummRp::_game.dataXorKey);
			if (ScummRp::_options & ScummRp::OPT_IMPORT)
			{
				ScummTr::_explore(*disk, ScummRp::ACT_IMPORT, text);
			}
			else
			{
				if (ScummTr::_exportWithPadding)
				{
					ScummIO::setQuiet(true);
					ScummTr::_explore(*disk, ScummTr::ACT_RSCNAMELIMITS, text);
					ScummIO::setQuiet(false);
				}
				ScummTr::_explore(*disk, ScummRp::ACT_EXPORT, text);
			}
			ScummRp::_mergeTmpIndex();
		}
	}
	ScummRp::_updateMainIndex();

	if (ScummRp::_options & ScummRp::OPT_IMPORT)
	{
		ScummIO::setQuiet(true);
		ScummRp::_exploreIndex<ScummRp::ACT_SAVE>(*index);
		ScummIO::setQuiet(false);
	}
}
#ifdef SCUMMTR_HAS_GOOD_GCC_DIAGNOSTIC_PRAGMA_FEATURES
#pragma GCC diagnostic pop
#endif

Text::Charset ScummTr::_selectCharset()
{
	if (ScummRp::_game.version > 2)
		return Text::CHS_V3_WIN1252;

	if (strcmp(ScummTr::_paramLanguage, "en") == 0)
		return Text::CHS_V1EN;
	if (strcmp(ScummTr::_paramLanguage, "de") == 0)
		return Text::CHS_V1DE;
	if (strcmp(ScummTr::_paramLanguage, "it") == 0)
		return Text::CHS_V1IT;
	if (strcmp(ScummTr::_paramLanguage, "fr") == 0)
		return Text::CHS_V1FR;

	ScummIO::warning(xsprintf("Unknown language code %s", ScummTr::_paramLanguage));

	return Text::CHS_NULL;
}

void ScummTr::setRscNameMaxLengh(ScummTr::RscType t, int32 id, int32 l)
{
	if (((1 << t) & ScummTr::_paddedRsc) == 0)
		return;

	if (id < 0)
	{
		if (ScummTr::_maxPadding && ScummTr::_rscNameLimits.anyRsc[t] < l)
			ScummTr::_rscNameLimits.anyRsc[t] = l;
	}
	else
	{
		if ((int32)ScummTr::_rscNameLimits.rsc[t].size() <= id)
			ScummTr::_rscNameLimits.rsc[t].resize(id + 1, 0);

		if (ScummTr::_rscNameLimits.rsc[t][id] < l)
			ScummTr::_rscNameLimits.rsc[t][id] = l;
	}
}

int32 ScummTr::getRscNameMaxLengh(ScummTr::RscType t, int32 id)
{
	if (((1 << t) & ScummTr::_paddedRsc) == 0)
		return 0;

	if (id < 0 || ScummTr::_rscNameLimits.anyRsc[t] >= ScummTr::_rscNameLimits.rsc[t][id])
		return ScummTr::_rscNameLimits.anyRsc[t];

	return ScummTr::_rscNameLimits.rsc[t][id];
}

int ScummTr::main(int argc, const char **argv)
{
	int g;

	ScummTr::_getOptions(argc, argv, ScummTr::_trParameters);
	ScummIO::setInfoSlots(ScummRp::_infoSlots);
	ScummIO::info(INF_GLOBAL, xsprintf("%s %s (build %s) by %s", ScummTr::NAME, ScummTr::VERSION, SCUMMTR_BUILD_DATE, ScummTr::AUTHOR));
	ScummIO::info(INF_GLOBAL, "");

	if (ScummRp::_options & ScummRp::OPT_LIST)
	{
		ScummTr::_listGames();
		return 0;
	}

	if (ScummTr::_invalidOptions())
	{
		ScummTr::_usage();
		return 0;
	}

	ScummTr::_paddedRsc = 0;
	for (int i = 0; ScummTr::_paramPaddedRsc[i] != '\0'; ++i)
	{
		switch (ScummTr::_paramPaddedRsc[i])
		{
		case 'a':
			ScummTr::_paddedRsc |= 1 << ScummTr::RSCT_ACTOR;
			break;
		case 'o':
			ScummTr::_paddedRsc |= 1 << ScummTr::RSCT_OBJECT;
			break;
		case 'v':
			ScummTr::_paddedRsc |= 1 << ScummTr::RSCT_VERB;
			break;
		default:
			ScummIO::fatal("-a/-A values must be: a, o, v");
			break;
		}
	}

	g = ScummRp::_findGameDef(ScummRp::_paramGameId);
	if (g == -1)
	{
		ScummRp::_listGames();
		return 0;
	}

	ScummRp::_game = ScummRp::_gameDef[g];

#ifndef SCUMMRP_OK_TO_CORRUPT_MANIACV2
	if (ScummRp::_game.version == 2 && ScummRp::_game.id == GID_MANIAC && ScummRp::_options & ScummRp::OPT_IMPORT)
		ScummIO::fatal("Modifying Maniac Mansion V2 is known to corrupt it");
#endif

	if (ScummRp::_options & ScummRp::OPT_IMPORT)
	{
		ScummRp::_fileOptions |= BlocksFile::BFOPT_BACKUP;
	}
	else
	{
		ScummRp::_fileOptions |= BlocksFile::BFOPT_READONLY;
		ScummTr::_textOptions |= Text::TXT_OUT;
	}

	if (ScummTr::_textOptions & Text::TXT_BINARY)
		ScummIO::warning("the -b option doesn't work reliably with all games");

	if (ScummRp::_game.version < 4)
		ScummTr::_processGameFilesV123();
	else
		ScummTr::_processGameFilesV4567();

	ScummRp::_backupSystem.applyChanges();

	return 0;
}

bool ScummTr::_readOption(const char *arg, char *pendingParams)
{
	char c;

	if (arg[0] != '-')
		return true;

	for (int i = 1; (c = arg[i]) != '\0'; i++)
	{
		switch (c)
		{
		case '-':
			return false;
		case 'h':
			ScummTr::_textOptions |= Text::TXT_HEADER;
			break;
		case 'H':
			ScummTr::_textOptions |= Text::TXT_HEXA;
			break;
		case 'I':
			ScummTr::_textOptions |= Text::TXT_OPCODE;
			break;
		case 'A':
			ScummTr::_exportWithPadding = true;
			ScummTr::_maxPadding = true;
			ScummRp::_queueParam(pendingParams, c);
			break;
		case 'a':
			ScummTr::_exportWithPadding = true;
			ScummTr::_maxPadding = false;
			ScummRp::_queueParam(pendingParams, c);
			break;
		case 'w':
			ScummTr::_textOptions |= Text::TXT_CRLF;
			break;
		case 'c':
			ScummTr::_textOptions |= Text::TXT_CHARSET1252;
			break;
		case 'b':
			ScummTr::_textOptions |= Text::TXT_BINARY;
			break;
		case 'o':
			ScummRp::_options |= ScummRp::OPT_EXPORT;
			break;
		case 'i':
			ScummRp::_options |= ScummRp::OPT_IMPORT;
			break;
		case 's':
			ScummRp::_fileOptions = BlocksFile::BFOPT_NULL;
			break;
		case 'm':
			ScummRp::_fileOptions &= ~BlocksFile::BFOPT_AUTO;
			ScummRp::_fileOptions |= BlocksFile::BFOPT_RAM;
			break;
		case 'n':
			ScummTr::_textOptions |= Text::TXT_NO_COMMENT;
			break;
		case 'O':
			ScummRp::_fileOptions &= ~BlocksFile::BFOPT_AUTO;
			ScummRp::_fileOptions |= BlocksFile::BFOPT_SEQFILE;
			break;
		case 'L':
			ScummRp::_options |= ScummRp::OPT_LIST;
			return false;
		case 'q':
			ScummRp::_infoSlots = INF_NULL; // intentionally not definitive
			break;
		case 'r':
			ScummTr::_textOptions |= Text::TXT_RAW;
			break;
		case 'v':
			ScummRp::_infoSlots |= INF_DETAIL;
			break;
// 		case 'V':
// 			ScummRp::_infoSlots |= INF_LISTING;
// 			break;
		case 'g':
			ScummRp::_options |= ScummRp::OPT_GAME_FILES;
			ScummRp::_queueParam(pendingParams, c);
			break;
		case 'l':
		case 'p':
		case 'f':
			ScummRp::_queueParam(pendingParams, c);
			break;
		default:
			ScummIO::fatal(xsprintf("Unrecognized \"-%c\" option", c));
			break;
		}
	}

	return true;
}

void ScummTr::_getOptions(int argc, const char **argv, const ScummRp::Parameter *params)
{
	char pendingParams[MAX_PARAMS + 1];

	pendingParams[0] = '\0';
	for (int i = 1; i < argc && argv[i] && ScummTr::_readOption(argv[i], pendingParams); ++i)
	{
		for (int j = 0; pendingParams[j] != '\0'; pendingParams[j++] = '\0')
		{
			for (int k = 0; params[k].c != '\0'; ++k)
			{
				if (params[k].c != pendingParams[j])
					continue;

				++i;
				if (i < argc && argv[i])
				{
					strncpy(params[k].value, argv[i], params[k].maxSize - 1);
					params[k].value[params[k].maxSize - 1] = '\0';
#if defined(_WIN32) || defined(__MSDOS__)
					if (params[k].isPath)
						for (char *p = strchr(params[k].value, '\\'); p; p = strchr(params[k].value, '\\'))
							*p++ = '/';
#endif
				}
				break;
			}
		}
	}
}

bool ScummTr::_invalidOptions()
{
	static const uint32 exclusive[] = { ScummRp::OPT_IMPORT | ScummRp::OPT_EXPORT, 0 };
	static const uint32 mandatory[] = { ScummRp::OPT_IMPORT | ScummRp::OPT_EXPORT, ScummRp::OPT_GAME_FILES, 0 };

	for (int i = 0; mandatory[i] != 0; ++i)
		if (!(ScummRp::_options & mandatory[i]))
			return true;

	for (int i = 0; exclusive[i] != 0; ++i)
		if ((ScummRp::_options & exclusive[i]) == exclusive[i])
			return true;

	return false;
}

void ScummTr::_usage()
{
	std::cout << "options:\n\n";
	std::cout << " -i         " << "import text into the game files (input)\n";
	std::cout << " -o         " << "export text from the game files (output)\n";
	std::cout << " -L         " << "list supported games\n\n";
	std::cout << " -a [aov]   " << "protect actors/objects/verbs for safer renames\n";
	std::cout << " -A [aov]   " << "same as -a, but with extra safety\n";
	std::cout << " -b         " << "binary mode (may not work with all games)\n";
	std::cout << " -c         " << "convert some Western European characters to Windows-1252\n";
	std::cout << " -f path    " << "path to the text file (default: " << ScummTr::_paramTextFile << ")\n";
	std::cout << " -g gameid  " << "select a game (as given by -L)\n";
	std::cout << " -h         " << "include SCUMM script context before each line\n";
	std::cout << " -H         " << "use hexadecimal values for escape sequences\n";
	std::cout << " -I         " << "include SCUMM instruction opcode before each line\n";
	std::cout << " -l xx      " << "language for -c (V1/V2 games only): en, de, it, fr\n";
// 	std::cout << " -m         " << "work in memory (whole game files are loaded in RAM)\n";
	std::cout << " -n         " << "never output any \"ScummTR note\" comment\n";
// 	std::cout << " -O         " << "optimize for sequential access (with -i)\n";
	std::cout << " -p path    " << "path to the game (default: current directory)\n";
	std::cout << " -q         " << "quiet mode\n";
	std::cout << " -r         " << "raw text (preserve original text encoding)\n";
// 	std::cout << " -s         " << "slow mode (disable automatic -m or -O)\n";
	std::cout << " -v         " << "verbose mode\n";
// 	std::cout << " -V         " << "more verbose (lists blocks)\n";
	std::cout << " -w         " << "use Windows CRLF newline characters\n\n";

	std::cout << "Examples:\n";
	std::cout << "scummtr -g zakv2    -l de -c -wh -A aov -of zak_v2_de.txt\n";
	std::cout << "scummtr -g monkeycd       -c -wh -A aov -of mi1_cd_en.txt\n";
	std::cout << "scummtr -g atlantis       -r -wh -A aov -of atlantis_jpn_shift_jis.txt\n\n";

	printCommonDisclaimer();
}
